### ISSUE TITLE
Why is withSyslog return type restricted to ()?

### ISSUE THREAD

**Issue by qrilka** [peti/hsyslog#15]
E.g. it results in the current `MonadLogger` instance from `monad-logger-syslog` to use `withSyslog` for every logging entry while in this package haddocks you write

> multiple calls to this function will interfere with each other in unpredictable ways

At the moment I don't see any obstacles to change `()` to `a`. Do I miss something maybe?

**Comment by peti** [comment]
I don't think there's any technical reason to force `()`. I made that choice mostly to guarantee that the logging function provided in the `withSyslog` scope cannot be leaked. Not sure whether that's a compelling reason. Maybe there's better alternatives to accomplish the same thing? Maybe it's good enough to trust users of the code not to shoot themselves in the foot?

**Comment by adinapoli** [comment]
Hey @peti ,

I know I'm a bit late to the party, but this morning I was working on a syslog-based [Katip](https://github.com/Soostone/katip) Scribe, and wondered the very same thing.

 > I made that choice mostly to guarantee that the logging function provided in the withSyslog scope cannot be leaked.

Isn't this a "false" problem? A user can simply use an `MVar` (like I did) and write something like this:

```
--------------------------------------------------------------------------------
mkSyslogScribe :: Severity -> IO Scribe
mkSyslogScribe sev = do
  syslogHandler <- newEmptyMVar
  withSyslog defaultConfig (\s -> putMVar syslogHandler s >> return ())
  return $ Scribe $ \ i@Item{..} -> do
      when (permitItem sev i) $ do
        withMVar syslogHandler $ \syslog -> syslog USER Error "lol"
```

As you can see, Katip's API forces me to return a `Scribe`, which is a sort of callback function to be called every time we have a logging item to call. As I didn't want to call `withSyslog` every time in the callback, I resorted to an MVar.

On the topic, can I ask you the implications of calling this function (`mkSyslogScribe`) in this case, multiple times, thus calling `withSyslog` more than once? `multiple calls to this function will interfere with each other in unpredictable ways` is a bit vague and I would love to learn more. Any idea where I could look at?

> Not sure whether that's a compelling reason. Maybe there's better alternatives to accomplish the same thing?

Hard to say when dealing with FFI stuff, maybe we should just wait for linear types to land in GHC 😉 

Thanks for writing this library!

**Comment by peti** [comment]
I just released version of the library, which comes comes with several improvements and simplifications to the API. One of the things I changed is that [withSyslog](http://hackage.haskell.org/package/hsyslog-5/docs/System-Posix-Syslog.html#v:withSyslog) no longer attempts to lock `syslog()` users into its function scope. The library even offers sensible versions of `openlog()` and `closelog()` so that you folks can implement whatever flow-control your application requires efficiently. I hope this is an improvement. Let me know if you have any feedback.
